[ENHANCEMENT][REFACTOR] SDK: allow to remove settings

frascuchon opened the pull request:
# Description
<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR allows users to change all the settings before creating the dataset, and metadata and vector, when the dataset is created. 

I included the `remove` method instead of a new `settings.add(override=True)` since this change requires a lot of internal refactor with the current design.

An example of how to use could be to change inferred settings from hub before create the final dataset

```python
settings = rg.Settings.from_hub("google/frames-benchmark")

settings.fields.remove("answer")

for field in settings.fields:
    if field.name.startswith("wiki"):
        settings.fields.remove(field)
        settings.metadata.add(rg.TermsMetadataProperty(field.name))

settings.questions.add(rg.TextQuestion(name="answer", title="Answer"))

dataset = rg.Dataset.from_hub("google/frames-benchmark", settings=settings)
```

Or adding new metadata or vector settings when the dataset is created:
```python
dataset = client.dataset("my-dataset")

dataset.metadata.add([rg.TermsMetadata(name="split")])
dataset.update() # this line sends the change to the argilla server

```

## Updated

You can remove a property by name:
```python
settings.fields.remove("text")
```

or by property instance
```python
for field in settings.fields:
   if field.name.startwith("wiki"):
      settings.fields.remove(field)
````

And you can override existing properties to change the property type, using the new `settings.add` method:
```python
settings = Settings.from_hub(...)

# change some settings definitions before create the dataset
settings.add(rg.TermsMetadata("wikipedia_1")) # this will remove the existing wikipedia_1 property and will create a new terms metadata one
```

**Type of change**
<!--  Please delete options that are not relevant. Remember to title the PR according to the type of change  -->

- Refactor (change restructuring the codebase without changing functionality)
- Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**
<!--  Please add some reference about how your feature has been tested.  -->

**Checklist**
<!--  Please go over the list and make sure you've taken everything into account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

frascuchon committed:
feat: Allow to remove settings

frascuchon committed:
chore: Add tests

frascuchon committed:
chore: Set proper attribute ref

frascuchon committed:
Apply suggestions from code review

burtenshaw commented:
Just so I understand, there is still a singular way to remove settings? And this is just an example of iteration:

```python
for field in settings.fields:
    if field.name.startswith("wiki"):
        settings.fields.remove(field)
        settings.metadata.add(rg.TermsMetadataProperty(field.name))
```

for example, fields can be deleted by their name:

```
settings.fields.remove("wiki_field")
```

burtenshaw committed:
Feature/add-property-in-settings-resource (#5587)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR implements a method to override settings properties in the
settings resource:

```python
settings = rg.Settings.from_hub("google/frames-benchmark")

settings.add(rg.TextQuestion(name="answer", title="Answer"))

dataset = rg.Dataset.from_hub("google/frames-benchmark", settings=settings, name=f"frames-benchmark-{uuid4()}")
```

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

burtenshaw commented:
@frascuchon Should we put all the add and remove options in the description and the get feedback on the API?

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

burtenshaw commented:
@frascuchon We haven't got feedback on this. I think we should move forward with the current implementation. Or ask again.

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

frascuchon committed:
chore: Adding tests

frascuchon committed:
docs: Update docs reference

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

frascuchon committed:
fix: Using base classes in conditions

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw commented on argilla/src/argilla/settings/_resource.py:
```diff
@@ -285,6 +286,43 @@ def from_hub(
+                    else:
+                        raise SettingsError(message)
+
+        if isinstance(property, FieldBase):
+            self.fields.add(property)
+        elif isinstance(property, QuestionPropertyBase):
+            self.questions.add(property)
+        elif isinstance(property, VectorField):
+            self.vectors.add(property)
+        elif isinstance(property, MetadataPropertyBase):
+            self.metadata.add(property)
```
Ok. This looks good.

burtenshaw commented:
@frascuchon This looks good to me.

sorry for the delay.

frascuchon committed:
Merge branch 'develop' into refactor/allow-remove-settings
